src/Raven.Server/Utils/DirectoryExecUtils.cs

```diff
@@ -0,0 +1,152 @@
+                return;
+
+            var directoryParameters = new DirectoryParameters
+            {
+                OnCreateDirectoryExec = config.OnCreateDirectoryExec,
+                OnCreateDirectoryExecArguments = config.OnCreateDirectoryExecArguments,
+                OnCreateDirectoryExecTimeout = config.OnCreateDirectoryExecTimeout.AsTimeSpan,
+                DatabaseName = databaseName,
+                Type = envType
+            };
+
+            void OnDirectory(StorageEnvironmentOptions internalOptions)
+            {
+                OnCreateDirectory(internalOptions, directoryParameters, logger);
+            }
+
+            options.OnCreateDirectory += OnDirectory;
+        }
+
+        public static void OnCreateDirectory(StorageEnvironmentOptions options, DirectoryParameters parameters, Logger log)
+        {
+            Process process = null;
+            try
+            {
+                var journalPath = string.Empty;
+                if (options is StorageEnvironmentOptions.DirectoryStorageEnvironmentOptions dirOptions)
+                    journalPath = dirOptions.JournalPath.FullPath;
+
+                var userArgs = parameters.OnCreateDirectoryExecArguments ?? string.Empty;
+                var args = $"{userArgs} {parameters.Type} {parameters.DatabaseName} " +
+                           $"{CommandLineArgumentEscaper.EscapeSingleArg(options.BasePath.ToString())} " +
+                           $"{CommandLineArgumentEscaper.EscapeSingleArg(options.TempPath.ToString())} " +
+                           $"{CommandLineArgumentEscaper.EscapeSingleArg(journalPath)}";
+
+                process = new Process
+                {
+                    StartInfo = new ProcessStartInfo
+                    {
+                        FileName = parameters.OnCreateDirectoryExec,
+                        Arguments = args,
+                        UseShellExecute = false,
+                        RedirectStandardOutput = true,
+                        RedirectStandardError = true,
+                        CreateNoWindow = true
+                    }
+                };
+                
+                var sw = Stopwatch.StartNew();
+
+                try
+                {
+                    process.Start();
+                }
+                catch (Exception e)
+                {
+                    throw new InvalidOperationException($"Unable to execute '{parameters.OnCreateDirectoryExec} {args}'. Failed to start process.", e);
+                }
+
+                var readStdOut = process.StandardOutput.ReadToEndAsync();
+                var readErrors = process.StandardError.ReadToEndAsync();
+
+                string GetStdError()
+                {
+                    try
+                    {
+                        return readErrors.Result;
+                    }
+                    catch
+                    {
+                        return "Unable to get stderr";
```

Comment:
No need to include the actual exception? Same below

Reply:
Yes, Igal will finish his extension point soon and we'll write tests for all four invocations of processes (directory, metrics, master key, certificate)